File: plugins/agent365/skills/purview-dlp-integration/assets/wiring-snippet.py
# ─────────────────────────────────────────────────────────────────────────────
# WIRING SNIPPET (PYTHON) — how to add the two DLP gates to an existing A365 agent.
#
# This is a REFERENCE, not an importable module. Apply these MINIMAL edits to
# response-returning handlers such as AgentInterface.process_user_message.
# The caller must send the returned text, including block/withhold messages.
# Do not paste this snippet into a host on_message callback: those callbacks must
# send activities explicitly via context.send_activity before returning.
# Notification/email handlers can use this pattern only if their caller sends the returned text.
# ─────────────────────────────────────────────────────────────────────────────

# 1) Import the guard at the top of your agent file:
from purview import purview_guard  # noqa: E402

# 2) Inside your message handler you already receive the A365 Authorization handler, the
#    auth-handler name, and the TurnContext. (In the AgentInterface pattern the signature is
#    `process_user_message(self, message, auth, auth_handler_name, context)`.) Resolve the caller id:
#
#    from_prop = getattr(getattr(context, "activity", None), "from_property", None)
#    user_id = (getattr(from_prop, "aad_object_id", None) or getattr(from_prop, "id", None) or "") if from_prop else ""


# Optional helper for the user-facing block/withhold messages:
def _dlp_block_message(verdict, phase: str) -> str:
    if verdict.decision == "blocked":
        return (
            "🚫 I can't help with that — your request was blocked by your organization's data-loss-prevention policy."
            if phase == "input"
            else "🚫 My response was withheld by your organization's data-loss-prevention policy."
        )
    return (
        "🚫 I couldn't check this against your organization's data policy, so I've stopped the request. Please try again shortly."
        if phase == "input"
        else "🚫 I couldn't check my response against your organization's data policy, so I've withheld it."
    )


# 3) INPUT GATE — evaluate the prompt BEFORE calling the LLM:
if purview_guard.is_enabled:
    gate = await purview_guard.evaluate_prompt(auth, auth_handler_name, user_id, message, context)
    if gate.blocked:
        return _dlp_block_message(gate, "input")  # ← the LLM is never called

# 4) ... your existing LLM call, e.g.:
#    reply = await self._agent.run(session_id, message)

# 5) OUTPUT GATE — evaluate the model's answer BEFORE returning it:
if purview_guard.is_enabled and purview_guard.is_check_output:
    out_gate = await purview_guard.evaluate_response(auth, auth_handler_name, user_id, reply, context)
    if out_gate.blocked:
        return _dlp_block_message(out_gate, "output")

# 6) ... your existing `return reply`; the caller sends it via context.send_activity.

File: plugins/agent365/skills/purview-dlp-integration/assets/wiring-snippet.ts
// ─────────────────────────────────────────────────────────────────────────────
// WIRING SNIPPET — how to add the two DLP gates to an existing A365 agent.
//
// This is a REFERENCE, not a compiled file. Apply these MINIMAL edits to your
// AgentApplication's message handler (and any notification/email handler that
// calls the LLM). Everything else stays the same.
// ─────────────────────────────────────────────────────────────────────────────

// 1) Import the guard at the top of your agent file:
import { purviewGuard } from './purview.js';

// 2) Inside your message handler, `context` is the TurnContext and `this.authorization`
//    is the AgentApplication's authorization handler. Resolve the caller id once:
//
//    const userId = context.activity.from?.aadObjectId ?? '';
//    const userMessage = context.activity.text?.trim() ?? '';

// 3) INPUT GATE — evaluate the prompt BEFORE calling the LLM:
if (purviewGuard.isEnabled) {
  const gate = await purviewGuard.evaluatePrompt(this.authorization, userId, userMessage, context);
  if (gate.blocked) {
    await context.sendActivity(
      gate.decision === 'blocked'
        ? "🚫 I can't help with that — your request was blocked by your organization's data-loss-prevention policy."
        : "🚫 I couldn't check this against your organization's data policy, so I've stopped the request. Please try again shortly.",
    );
    return; // ← the LLM is never called
  }
}

// 4) ... your existing LLM call, e.g.:
//    const answer = await model.invoke(userMessage);

// 5) OUTPUT GATE — evaluate the model's answer BEFORE sending it:
if (purviewGuard.isEnabled && purviewGuard.isCheckOutput) {
  const outGate = await purviewGuard.evaluateResponse(this.authorization, userId, answer, context);
  if (outGate.blocked) {
    await context.sendActivity(
      outGate.decision === 'blocked'
        ? "🚫 My response was withheld by your organization's data-loss-prevention policy."
        : "🚫 I couldn't check my response against your organization's data policy, so I've withheld it.",
    );
    return;
  }
}

// 6) ... your existing `await context.sendActivity(answer);`
